### PR TITLE
Securing SP's APIs and dashboard artifacts

### PR DESCRIPTION
**Pull request by Niveathika92**
## Purpose
To secure Steam Processor APIs and dashboard artifacts.

## Goals
Add an interface that interacts with a userstore.
Add a REST API so that a user can login/logout using UI.
Add an interceptor to intercept requests to the APIs so that the users can be authenticated.

## Approach
1. Adding an high level interface for IdP interactions. 
2. Exposing IdPClient an OSGi-service
3. Adding request interceptors for all APIs other than authentication API (login/logout)
4. Using authenticate mechanism in the IdPClient osgi-service to authenticate users who sends requests to API.
5. Exposing an authentication API which can be used by the UI to login/logout users to the system.

## User stories
1. System should be able to initialize the IdPClient OSGi-service based on user configurations.
2. Interceptor should be able to use IdPClient OSGi-service to auhenticate users.
3. Users should be able to gain access to the system's APIs when using UI with the use of Auth API (login/logout)

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: components/org.wso2.carbon.msf4j.interceptor.common/pom.xml [hunk 1]
```diff
@@ -40,6 +40,10 @@
             <groupId>org.wso2.msf4j</groupId>
```
Comment (suhothayan) [inline components/org.wso2.carbon.msf4j.interceptor.common/pom.xml]: I think you should not use org.wso2.carbon.msf4j.interceptor.common
but use org.wso2.carbon.**analytics**.msf4j.interceptor.common

Comment (suhothayan) [reply components/org.wso2.carbon.msf4j.interceptor.common/pom.xml]: same for the packages

File: components/org.wso2.carbon.msf4j.interceptor.common/src/main/java/org/wso2/carbon/msf4j/interceptor/common/internal/ServiceComponent.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            service = IdPClient.class,
+            cardinality = ReferenceCardinality.MANDATORY,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unregisterIdP"
+    )
+    protected void registerIdP(IdPClient client) {
+        DataHolder.getInstance().setIdPClient(client);
+    }
+
+
```
Comment (suhothayan) [inline components/org.wso2.carbon.msf4j.interceptor.common/src/main/java/org/wso2/carbon/msf4j/interceptor/common/internal/ServiceComponent.java]: Remove unnecessary spaces

File: components/org.wso2.carbon.stream.processor.auth.rest.api/pom.xml [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            <artifactId>org.wso2.carbon.msf4j.interceptor.common</artifactId>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.wso2.maven.plugins</groupId>
+                <artifactId>swagger2msf4j-maven-plugin</artifactId>
+                <version>1.0-SNAPSHOT</version>
```
Comment (suhothayan) [inline components/org.wso2.carbon.stream.processor.auth.rest.api/pom.xml]: the versions should be moved to parent pom

Comment (suhothayan) [reply components/org.wso2.carbon.stream.processor.auth.rest.api/pom.xml]: For all the plugins too

File: components/org.wso2.carbon.msf4j.interceptor.common/pom.xml (hunk 1 above)
Comment (Niveathika92) [reply components/org.wso2.carbon.msf4j.interceptor.common/pom.xml]: Fixed in 497ede4

File: components/org.wso2.carbon.msf4j.interceptor.common/src/main/java/org/wso2/carbon/msf4j/interceptor/common/internal/ServiceComponent.java (hunk 2 above)
Comment (Niveathika92) [reply components/org.wso2.carbon.msf4j.interceptor.common/src/main/java/org/wso2/carbon/msf4j/interceptor/common/internal/ServiceComponent.java]: Fixed in 497ede4

File: components/org.wso2.carbon.stream.processor.auth.rest.api/pom.xml (hunk 3 above)
Comment (Niveathika92) [reply components/org.wso2.carbon.stream.processor.auth.rest.api/pom.xml]: Fixed in dbca42f

**Comment by Niveathika92** [comment]
Closing the PR since it was decided to move the IdP components and features to analytics-common

**Pull request closed without merging**